pkg/loki/loki.go

```diff
@@ -20,50 +23,158 @@ func init() {
+		// If an old instance existed, update it and move it to the new map.
+		if old, ok := l.instances[ic.Name]; ok {
+			err := old.ApplyConfig(ic)
+			if err != nil {
+				return err
+			}
+
+			delete(l.instances, ic.Name)
+			newInstances[ic.Name] = old
 			continue
 		}
 
-		r := prometheus.WrapRegistererWith(prometheus.Labels{"loki_config": ic.Name}, reg)
-		il := log.With(l, "loki_config", ic.Name)
-
-		p, err := promtail.New(config.Config{
-			ServerConfig:    server.Config{Disable: true},
-			ClientConfigs:   ic.ClientConfigs,
-			PositionsConfig: ic.PositionsConfig,
-			ScrapeConfig:    ic.ScrapeConfig,
-			TargetConfig:    ic.TargetConfig,
-		}, false, promtail.WithLogger(il), promtail.WithRegisterer(r))
+		inst, err := NewInstance(l.reg, ic, l.l)
 		if err != nil {
-			return nil, err
+			return fmt.Errorf("unable to apply config for %s: %w", ic.Name, err)
 		}
+		newInstances[ic.Name] = inst
+	}
 
-		promtails = append(promtails, p)
+	// Any remaining promtail in l.instances has been removed from the new
+	// config. Stop them before replacing the map.
+	for _, i := range l.instances {
+		i.Stop()
 	}
+	l.instances = newInstances
 
-	return &Loki{promtails: promtails}, nil
+	return nil
 }
 
 func (l *Loki) Stop() {
-	for _, p := range l.promtails {
-		p.Shutdown()
+	l.mut.Lock()
+	defer l.mut.Unlock()
+
+	for _, i := range l.instances {
+		i.Stop()
+	}
+}
+
+// Instance is an individual Loki instance.
+type Instance struct {
+	mut sync.Mutex
+
+	cfg *InstanceConfig
+	log log.Logger
+	reg *util.Unregisterer
+
+	promtail *promtail.Promtail
+}
+
+// NewInstance creates and starts a Loki instance.
+func NewInstance(reg prometheus.Registerer, c *InstanceConfig, l log.Logger) (*Instance, error) {
+	instReg := prometheus.WrapRegistererWith(prometheus.Labels{"loki_config": c.Name}, reg)
+
+	inst := Instance{
+		reg: util.WrapWithUnregisterer(instReg),
+		log: log.With(l, "loki_config", c.Name),
+	}
+	if err := inst.ApplyConfig(c); err != nil {
+		return nil, err
+	}
+	return &inst, nil
+}
+
+// ApplyConfig will apply a new InstanceConfig. If the config hasn't changed,
+// then nothing will happen, otherwise the old Promtail will be stopped and
+// then replaced with a new one.
+func (i *Instance) ApplyConfig(c *InstanceConfig) error {
+	i.mut.Lock()
+	defer i.mut.Unlock()
+
+	// No-op if the configs haven't changed.
+	if cmp.Equal(c, i.cfg) {
+		level.Debug(i.log).Log("msg", "instance config hasn't changed, not recreating Promtail")
+		return nil
+	}
+	i.cfg = c
+
+	if i.promtail != nil {
+		i.promtail.Shutdown()
+		i.promtail = nil
+	}
+
+	// Unregister all existing metrics before trying to create a new instance.
+	if !i.reg.UnregisterAll() {
+		// If UnregisterAll fails, we need to abort, otherwise the new promtail
+		// would try to re-register an existing metric and might panic.
+		return fmt.Errorf("failed to unregister all metrics from previous promtail. THIS IS A BUG!")
```

Comment:
Should some information about `i.reg` be included here?

Reply:
Hmm, maybe, though there's really not any information we can get from it

Reply:
Yeah, it just holds interfaces. I'll at least add something for which instance it was.

Reply:
Yeah, it just holds interfaces so there's nothing we could include here. The outer ApplyConfig will say which instance it was for, at least.